### PR TITLE
Human-readable and actionable SQL error messages

### PR DESCRIPTION
**Pull request by abhishekagarwal87**
### Description

This PR does two things
 - It adds the capability to surface missing features in SQL to users - The calcite planner will explore through multiple rules to convert a logical SQL query to a druid native query. Some rules change the shape of the query itself, optimize it and some rules are responsible for translating the query into a druid native query. These are DruidQueryRule, DruidOuterQueryRule, DruidJoinRule, DruidUnionDataSourceRule, DruidUnionRule etc. These rules will look at SQL and will do the necessary transformation. But if the rule can't transform the query, it returns back the control to the calcite planner without recording why was it not able to transform. E.g. there is a join query with a non-equal join condition. DruidJoinRule will look at the condition, see that it is not supported, and return back the control. The reason can be that a query can be planned in many different ways so if one rule can't parse it, the query may still be parseable by other rules. In this PR, we are intercepting these gaps and passing them back to the user if the query could not be planned at all. 
 - The said capability has been used to generate actionable errors for some common unsupported SQL features. However, not all possible errors are covered and we can keep adding more in the future. 
 
The code changes themselves are not intrusive and minimal. I have modified the `PlannerContext` to record any planning error and then use it when generating a user-facing exception. 

Update 22-Nov-2021
I added a few more enhancements. In many places, we are throwing IAE, ISE exception during query planning e.g. when the query includes max aggregation on a string column, we throw ISE. This is caught by calcite, wrapped inside RuntimeException, and thrown up. This halts the processing and the error that uses gets to see in the console is not helpful at all since it is buried inside the cause chain. In this patch, I have replaced these exceptions with `UnsupportedSQLQueryException`. In DruidPlanner, the code tries to find the root cause that is of the type CannotPlanException. If such a cause is found, we extract the error and use that as the top-level error message. The reason behind using `UnsupportedSQLQueryException` is that we can be sure that this exception class will be populated with user-friendly error message. For any other CannotPlanningException, we return a generic error message. 

##### Key changed/added classes in this PR
 * `DruidPlanner`
 * `Druid*Rule`
 * `DruidUnionDataSourceRel`

<hr>

This PR has:
- [ ] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by gianm** [comment]
The technique here is cool and an interesting way around the fact that these rules can be "exploratory" and therefore they can't simply throw exceptions.

But, is there a potential for bizarre errors that relate to rules or query types that were explored, but didn't work out, and appear like nonsense to the user? We'd want to make sure that every time `setPlanningError` is called, it's something so fundamental that the overall query can't be handled by a different planner path. The idea is that we want to avoid a situation where there are two paths A and B, and each has an error, and the error from A is reported but doesn't make sense to the user because it is sort of path-A-specific.

In other words: we should make sure the error messages always refer to things that the user actually wrote in their query, and that we know we don't support. I suggest adding something about this to the javadocs for the method so people don't add new ones that end up being confusing.

Although, that also makes me think: if these errors are going to be really unrecoverable, maybe we _should_ actually make them regular exceptions that halt planning immediately…?

**Comment by abhishekagarwal87** [comment]
> The technique here is cool and an interesting way around the fact that these rules can be "exploratory" and therefore they can't simply throw exceptions.
> 
> But, is there a potential for bizarre errors that relate to rules or query types that were explored, but didn't work out, and appear like nonsense to the user? We'd want to make sure that every time `setPlanningError` is called, it's something so fundamental that the overall query can't be handled by a different planner path. The idea is that we want to avoid a situation where there are two paths A and B, and each has an error, and the error from A is reported but doesn't make sense to the user because it is sort of path-A-specific.

Yes. It is possible. Here is one such example query
```
SELECT dim2, COUNT(distinct dim1), COUNT(distinct dim2) FROM druid.foo GROUP BY dim2
```
This query should ideally throw an error that suggests that multiple exact distinct count are not supported unless you set a custom flag. However, the error that the user gets is below
```
Only equal conditions are supported in joins
```
This would, of course, look bizarre to end users since they have not put a join in the query and it wouldn't be clear to them that a query with multiple exact distinct counts could be planned using a join. I am thinking of some ways of how we can make this experience better
- Be smarter about setting these errors and improve on a case-by-case basis. E.g. in the above example, I might check if the join condition is of type `is-distinct-from` and if so, I add an additional note to the error message such as below
```
if you are using multiple exact distinct counts, please make sure to turn flag X on.
```
- Use error codes alongside error messages and document any such gotchas in our user documentation. So when user searches for an error code, they can get more details about the error message and possible query shapes that could fail with that error. 

None of these are perfect solutions but will minimize the confusion to a good extent. Your suggestion about being conservative about setting the planning error makes sense so long we cover most of the common SQL errors that our users run into. 
> 
> In other words: we should make sure the error messages always refer to things that the user actually wrote in their query, and that we know we don't support. I suggest adding something about this to the javadocs for the method so people don't add new ones that end up being confusing.

Yes. SGTM. 
 
> Although, that also makes me think: if these errors are going to be really unrecoverable, maybe we _should_ actually make them regular exceptions that halt planning immediately…?

I thought about it but the risk of a false positive (falsely declaring a query as unsupported) becomes very high with that approach. In the current scheme, setting the planning error is a no-op if the query does get planned eventually. If we are halting the planning immediately, we may end up killing genuine queries. The current approach, I feel, is safer in that way. And in the future, a developer, trying to add an error message, will not have to struggle with the dilemma of breaking some genuine query in exchange for an informative error message.

**Comment by cryptoe** [comment]
IMHO We can change the messaging a bit. Rather than calling it an error, we can say "QueryDiagnosis".
`We tried to change the query shape to an intermediate "Union/Join/......" query but could not due to "reason".`

Also, I think for the first cut `setPlannerError` should append errors rather than overwrite the error. This way, we can show all the possible "error" paths to the end-user. This way we are safeguarding against leading the user down the incorrect diagnosis path. Consider the scenario where the user wrote a 1000 line SQL query :) . Debugging such a query with an incorrect error message may not be the most pleasant experience :)

If we want to show the error I echo Gian's thought:

> But, is there a potential for bizarre errors that relate to rules or query types that were explored, but didn't work out, and appear like nonsense to the user? We'd want to make sure that every time `setPlanningError` is called, it's something so fundamental that the overall query can't be handled by a different planner path. The idea is that we want to avoid a situation where there are two paths A and B, and each has an error, and the error from A is reported but doesn't make sense to the user because it is sort of path-A-specific.

**Review by jihoonson — COMMENTED** [review]
@abhishekagarwal87 thanks for the PR. The direction of this PR makes sense to me. The goal in that direction should be helping end-users understand query errors and fix their queries by themselves if possible. In that regard, the error message at least should not mislead users, and thus +1 for error message to start with `Possible error`, `Potential cause`, `Query diagnosis` or anything that is most obvious that the error might be wrong. 

> Use error codes alongside error messages and document any such gotchas in our user documentation. So when user searches for an error code, they can get more details about the error message and possible query shapes that could fail with that error.

This sounds interesting. Perhaps we can add this doc in follow-ups of this PR.

> Also, I think for the first cut setPlannerError should append errors rather than overwrite the error. This way, we can show all the possible "error" paths to the end-user. This way we are safeguarding against leading the user down the incorrect diagnosis path. Consider the scenario where the user wrote a 1000 line SQL query :) . Debugging such a query with an incorrect error message may not be the most pleasant experience :)

Similar to the case in Gian's comment, I suppose this could be a problem in a case like this: the query planner tried 2 different transforms. Both transformed queries are not supported by druid, but there is a workaround for one transformed query, such as turning on some features which are disabled by default, and nothing for another. However, the user could see only one error that has no workaround because `planningError` was overwritten. @abhishekagarwal87 does this case make sense? If so, it seems more reasonable to show users all potential errors.

> Also, I think for the first cut setPlannerError should append errors rather than overwrite the error. This way, we can show all the possible "error" paths to the end-user. This way we are safeguarding against leading the user down the incorrect diagnosis path. Consider the scenario where the user wrote a 1000 line SQL query :) . Debugging such a query with an incorrect error message may not be the most pleasant experience :)

This is a good call, but I think showing all potential errors won't help with debugging a 1000 line SQL query anyway since the query is too big. Ideally, we should be able to provide some context for each error where they are coming from in the query. I'm not sure how to do it though, maybe we can keep some context of the parent node while traversing the query plan tree.

**Comment by gianm** [comment]
> This is a good call, but I think showing all potential errors won't help with debugging a 1000 line SQL query anyway since the query is too big. Ideally, we should be able to provide some context for each error where they are coming from in the query. I'm not sure how to do it though, maybe we can keep some context of the parent node while traversing the query plan tree.

The best mechanism in Calcite for doing what you describe is post-parsing validation. It operates at an earlier stage: on the parse tree instead of the relational operator tree. Most of Calcite's builtin validations happen at that stage, and at that stage there is information available about line numbers, etc. Also it generally assumes that anything that makes it past the validator is a valid query, so I think that's why it has really "weird" errors from the planner when the planner can't plan something.

I added some validations in another patch that operate on that parse tree (although they don't use line number info): https://github.com/apache/druid/blob/61bc7edd163b447fb7ec7a645f074f1e6320f36b/sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java#L542-L577

Maybe we can use a similar technique for this stuff? i.e. whenever possible, look at the parse tree and identify certain patterns we cannot handle, before handing it off to the planner.

**Comment by abhishekagarwal87** [comment]
Surfacing the error in post-parsing validation can be done in some places e.g. in MaxSQLAggregator, druid can tell calcite that string operands are not supported. This can be done by returning a custom operator with supported types in `calciteFunction()` instead of returning a standard operator from the calcite library.  I do not know how would we do that at other places such as an unsupported join condition or union between incompatible data sources. 

It is good to see these ideas coming in for how to surface these errors in a very useful. There are two major areas of improvement I can see based on the discussion so far
 - one is the new error framework itself that is added in this PR. Here the debate is about showcasing one error or multiple errors. There is also the question of do we need this framework at all. I think we can try to minimize using it but in some places, plannerContext is the only way to surface the error. 
 - second is improving each error or scenario itself. By improving, I mean making it precise and as informative as possible. I believe this will be an ongoing exercise. In this PR, I replaced `ISE/IAE` with a custom exception to surface many errors. Going one step ahead and using calcite to surface these errors is even better as that is possible in some places. And we can continue to improve these errors and make them more precise in future improvements. I do not think it will be possible to do all of that in this PR itself. :) 

So here are the open questions 
 - retaining last planning error or last X number of planning errors - I am in the camp of former since I did want to see some real-world usage first before even trying to do the latter. since these are errors, we can always change them later on. That being said, I am not very strongly opposed to the idea of retaining multiple planning errors either. 
 - HTTP response code for an unplannable query - Right now, we are returning 500 but should we return 400 instead? In fact, we do return 400 as well for some planning errors. Should we treat CannotPlanException, not as an internal server exception but as a bad query request?
 - Delegating the operand validation to calcite in some of the aggregation functions - This I can do in this PR itself. 
 - scope of the initial PR - I have attempted to clarify that in the first few paragraphs. Happy to hear any other thoughts on it.

**Review by jihoonson — COMMENTED** [review]
> Maybe we can use a similar technique for this stuff? i.e. whenever possible, look at the parse tree and identify certain patterns we cannot handle, before handing it off to the planner.

As @abhishekagarwal87 said, I'm not sure either if this technique can be used to handle any kind of unsupported feature errors because, when identifying those patterns, we should also consider the possibility that the query is executable after the tree is transformed into another form. As this seems not much different than "query planning", it looks reasonable to do it as a part of query planning after post-parsing validation. Maybe we can add another validation phase when query planning fails after exhausting every possible candidate plan? The new validation phase can pick up the parse tree after planning fails and find any matching pattern of unsupported features. I think we can continue discussion on this matter without blocking this PR necessarily.

> - one is the new error framework itself that is added in this PR. Here the debate is about showcasing one error or multiple errors. There is also the question of do we need this framework at all. I think we can try to minimize using it but in some places, plannerContext is the only way to surface the error.

I think the new framework is certainly valuable as it opens a new opportunity for better error handling. I don't have a strong opinion on one error vs multiple errors.

> - second is improving each error or scenario itself. By improving, I mean making it precise and as informative as possible. I believe this will be an ongoing exercise. In this PR, I replaced ISE/IAE with a custom exception to surface many errors. Going one step ahead and using calcite to surface these errors is even better as that is possible in some places. And we can continue to improve these errors and make them more precise in future improvements. I do not think it will be possible to do all of that in this PR itself. :)

I think it makes sense as long as the new errors don't mislead users.

> - HTTP response code for an unplannable query - Right now, we are returning 500 but should we return 400 instead? In fact, we do return 400 as well for some planning errors. Should we treat CannotPlanException, not as an internal server exception but as a bad query request?

Good question. It makes sense to me to return 400 in this case. If you are going to make this change, please add "Release notes" and "incompatible" tags to this PR. It would be nice to add a "release note" section in the PR description about the incompatible change to help the release manager in the future.

Travis CI failure seems legit. Please check it. +1 after CI is fixed. 

Finally, I tagged this PR "Design Review" because this is an end-user facing feature. You won't need any more reviewers though since there are already enough committers involved.

**Comment by abhishekagarwal87** [comment]
Thanks @jihoonson 
Apart from addressing your comments, I am also going to re-word the errors a bit
e.g. "Possible error: only equal conditions are supported in join" --> "Possible error: SQL query requires a join with condition that is not supported. Currently, only equal conditions are supported"

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java
Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java]: I made the changes in `QueryLifecycle` to be consistent with logging here.

**Review by jihoonson — APPROVED** [review]
LGTM. Thanks @abhishekagarwal87!

**Pull request merged**
